Set use_blas64 to false to fix MKL.jl interaction numpy problem

fgerick opened the pull request:
same as the other pr, updated to latest master.

crstnbr commented:
Would be good to link to [the other PR](https://github.com/JuliaComputing/MKL.jl/pull/11).

aminya commented:
This builds now with the new version. I don't see any reason for not merging this.

crstnbr commented:
What precisely are the implications of `USE_BLAS64=false`? (apart form fixing the numpy interplay)

fgerick commented:
is it easy to include a flag in the build for the individual user to choose?

KristofferC commented on src/install.jl:
```diff
@@ -4,11 +4,15 @@ function replace_libblas(base_dir, name)
     lines[libblas_idx] = "const libblas_name = $(repr(name))"
     lines[liblapack_idx] = "const liblapack_name = $(repr(name))"
+    if useblas64_idx != nothing
+        lines[useblas64_idx] = "const USE_BLAS64 = false"
```
This will set it to use 32 bit integers even when reverting to normal sysimage? Anyway, I can fix it up.

ViralBShah replied:
Bump. @KristofferC Not sure I understand this - but it would be nice to get this PR merged.

crstnbr commented:
> is it easy to include a flag in the build for the individual user to choose?

It's probably fine to simply make it the default. But I think it's worth mentioning in the README.md.

aminya commented:
Some time ago I did some comparison between BLAS 64 and 32, and the performance gain of BLAS 64 is noticeable. I think BLAS 32 should not be the default behavior, but only be enabled by an option,

ViralBShah commented:
> Some time ago I did some comparison between BLAS 64 and 32, and the performance gain of BLAS 64 was noticeable. I think BLAS 32 should not be the default behavior, but only be enabled by an option,

It's a bit odd. Can you share the benchmarks?

ViralBShah commented:
Even if there is a performance issue, we probably still need to ship 32-bit as the default, given the number of crash complaints we receive.

ViralBShah commented:
What is holding this PR from being merged?

ViralBShah commented:
@stevengj If we merge this - I assume it will solve all the issues we are having on the python side.

stevengj commented:
Probably.

aminya commented:
@ViralBShah I created an issue to remind myself to do some benchmarks
https://github.com/JuliaComputing/MKL.jl/issues/50

fgerick commented:
I have added the option to build MKL.jl with the 64bit version of MKL by setting `ENV["USE_BLAS64"]=true` before building. Also, the install script reverts the alteration by the installation. I added a small statement into the README, but maybe someone wants to say something about performance issues...

fgerick commented:
I guess once this is merged the comment in the readme: "frequently encountered compatibility issues" is already deprecated...

fgerick commented:
Is there anything left to be done here? I'd be happy to see this move on and make MKL with Julia more end-user friendly. I'm still avoiding MKL because of [Arpack.jl](https://github.com/JuliaLinearAlgebra/Arpack.jl) but I'm hoping if MKL.jl has less struggles steps will be taken there also.

ViralBShah commented:
We need @andreasnoack to look at this - whenever he has a moment.

ViralBShah commented:
Failing on nightly.

andreasnoack commented:
@KristofferC Is the failure on nightly known?

ViralBShah commented:
https://travis-ci.org/github/JuliaComputing/MKL.jl/jobs/735062657#L262

```
ERROR: ArgumentError: Package Artifacts not found in current path:
- Run `import Pkg; Pkg.add("Artifacts")` to install the Artifacts package.
```

andreasnoack commented:
I suppose it's an issue with PackageCompiler on 1.6? I'll merge, as tests are passing on 1.5.

KristofferC commented:
How will this work with MKLSparse? Will we lose the ability to do Int64 sparse matmuls?

andreasnoack commented:
> How will this work with MKLSparse? Will we lose the ability to do Int64 sparse matmuls?

@fgerick Would you be able to test this?

KristofferC commented:
It seems MKL at compile time uses `BlasInt`: https://github.com/JuliaSparse/MKLSparse.jl/blob/6060cd08a2586b275c95aac650c379ff48130ef7/src/BLAS/level_2_3/generator.jl#L36

So I guess that means it would be restricted to `Int32` (and silently for `Int64` use the Julia native one)?

ViralBShah commented:
Yes, presumably. I think that is also the behaviour we need by default, otherwise it leads to all those crashes in matplotlib.

KristofferC commented:
Yeah, all I'm saying is that right now, if someone installs this it most likely effectively turns off MKLSparse since everyone is using the default `Int64` for their sparse matrices. And I guess many people that use MKLSparse uses MKL (and vice versa).

ViralBShah commented:
Oh I didn't quite understand that. Good to know. Is there a way to opt into the Int64 version?
